specs/merkle-tree.md

```diff
@@ -0,0 +1,158 @@
+
+### 4. There must be exactly one approval leaf or cancel leaf per chain
+Since Sphinx Merkle trees can contain an arbitrary set of leaves, it is possible to construct a single tree that contains `APPROVE` or `CANCEL` leaves that approve or cancel multiple separate deployments on a single chain. However, this leads to an ambiguous situation for the relayer where it is unclear which deployment should be approved or canceled.
+
+Additionally, The `SphinxModuleProxy` does not support both canceling an existing deployment and approving a new deployment in the same tree, therefore any tree that contains both a `CANCEL` and `APPROVE` leaf for a given chain would not be executable on that chain. So we impose the standard, that in a given Sphinx Merkle tree there should be exactly one `APPROVE` or `CANCEL` leaf per chain.
+
+If the `arbitraryChain` field is set to true, then we consider this approval leaf to apply to all chains. Therefore if `arbitraryChain` = true, then there should be exactly one `APPROVE` leaf in the entire tree and no `CANCEL` leaves.
+
+## Sphinx Merkle Tree Generation Standard Implementation
+We provide a standard implementation for generating Sphinx Merkle trees that adhere to the above invariants. This implementation is used by the official Sphinx Foundry plugin.
+
+### Input
+Accepts an object where the keys are canonical chain ids and the values are deployment data objects which contain all of the necessary info to assemble a SphinxMerkleTree. See the [inline documentation](TODO(end)) for exact details on the input data. The Sphinx Merkle tree generation function should be agnostic to the source of the transaction data so it can be used to assemble a tree based on transactions generated from any scripting framework (i.e Foundry, Hardhat Ignition, some arbitrary future framework).
+
+### Output
+A Sphinx Merkle tree that follows the above specification and invariants. See the [inline documetation](TODO(end)) for the exact details of the output data structure.
+
+## Function-Level Invariants
+We export three separate functions `makeSphinxLeaves`, `makeSphinxMerkleTreeFromLeaves`, and `makeSphinxMerkleTree`. We provide three separate functions for convenience and modularity to enable testing. However, for this specification and the below invariants we assume the use of the `makeSphinxMerkleTree`. We *only* recommend using that function when generating Sphinx Merkle trees in production and therefore only consider it to satisfy the invariants listed above.
+
+### It is impossible to specify multiple `NetworkDeploymentData` objects for a given network
+The Merkle tree generation logic relies on the fact that it is not possible to submit multiple `NetworkDeploymentData` objects to guarantee that there is exactly one `APPROVE` or `CANCEL` leaf ([4](#4-there-must-be-exactly-one-approval-leaf-or-cancel-leaf-per-chain-4)).
```

Comment:
I think this can be shortened to: This guarantees that there is exactly one `APPROVE` or `CANCEL` leaf ([4](#4-there-must-be-exactly-one-approval-leaf-or-cancel-leaf-per-chain-4)).